exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/EntityCache.kt

```diff
@@ -3,155 +3,203 @@ package org.jetbrains.exposed.dao
 import org.jetbrains.exposed.dao.id.EntityID
 import org.jetbrains.exposed.dao.id.IdTable
 import org.jetbrains.exposed.sql.*
-import org.jetbrains.exposed.sql.transactions.transactionScope
+import org.jetbrains.exposed.sql.transactions.ITransaction
 import java.util.*
-import java.util.concurrent.CopyOnWriteArrayList
-
-val Transaction.entityCache : EntityCache by transactionScope { EntityCache(this) }
 
 @Suppress("UNCHECKED_CAST")
-class EntityCache(private val transaction: Transaction) {
-    internal var flushingEntities by transactionScope { false }
-    val data = LinkedHashMap<IdTable<*>, MutableMap<Any, Entity<*>>>()
-    val inserts = LinkedHashMap<IdTable<*>, MutableList<Entity<*>>>()
-    val referrers = HashMap<EntityID<*>, MutableMap<Column<*>, SizedIterable<*>>>()
-
-    private fun getMap(f: EntityClass<*, *>) : MutableMap<Any, Entity<*>> = getMap(f.table)
-
-    private fun getMap(table: IdTable<*>) : MutableMap<Any, Entity<*>> = data.getOrPut(table) {
-        LinkedHashMap()
-    }
-
-    fun <ID: Any, R: Entity<ID>> getOrPutReferrers(sourceId: EntityID<*>, key: Column<*>, refs: ()-> SizedIterable<R>): SizedIterable<R> =
-            referrers.getOrPut(sourceId){ HashMap() }.getOrPut(key) { LazySizedCollection(refs()) } as SizedIterable<R>
-
-    fun <ID:Comparable<ID>, T: Entity<ID>> find(f: EntityClass<ID, T>, id: EntityID<ID>): T? = getMap(f)[id.value] as T? ?: inserts[f.table]?.firstOrNull { it.id == id } as? T
-
-    fun <ID:Comparable<ID>, T: Entity<ID>> findAll(f: EntityClass<ID, T>): Collection<T> = getMap(f).values as Collection<T>
-
-    fun <ID:Comparable<ID>, T: Entity<ID>> store(f: EntityClass<ID, T>, o: T) {
-        getMap(f)[o.id.value] = o
-    }
-
-    fun store(o: Entity<*>) {
-        getMap(o.klass.table)[o.id.value] = o
-    }
-
-    fun <ID:Comparable<ID>, T: Entity<ID>> remove(table: IdTable<ID>, o: T) {
-        getMap(table).remove(o.id.value)
-    }
-
-    fun <ID:Comparable<ID>, T: Entity<ID>> scheduleInsert(f: EntityClass<ID, T>, o: T) {
-        inserts.getOrPut(f.table) { arrayListOf() }.add(o as Entity<*>)
```

Comment:
it seems there is a kotlin bug that causes gc to reap the contents of lambda-based init function here. at the moment that the delegate (DaoTransaction) returned from this call, the `inserts` member would be empty, but a return reference to it would be valid (this was done as a differential test only and not committed.) changing to the new code, while ugly, solved this bug and allowed the remaining unit tests to pass. (You can set this line back and see the tests that don't pass and step out of this call to see the odd behavior)